bgzf.c

```diff
@@ -593,36 +595,67 @@ static int inflate_block(BGZF* fp, int block_length)
+static int inflate_gzip_block(BGZF *fp)
 {
-    int ret = Z_OK;
-    do
-    {
-        if ( !cached && fp->gz_stream->avail_out!=0 )
-        {
-            fp->gz_stream->avail_in = hread(fp->fp, fp->compressed_block, BGZF_BLOCK_SIZE);
-            if ( fp->gz_stream->avail_in<=0 ) return fp->gz_stream->avail_in;
-            if ( fp->gz_stream->avail_in==0 ) break;
+    // we will set this to true when we detect EOF, so we don't bang against the EOF more than once per call
+    int input_eof = 0;
+    
+    // write to the part of the output buffer after block_offset
+    fp->gz_stream->next_out = (Bytef*)fp->uncompressed_block + fp->block_offset;
+    fp->gz_stream->avail_out = BGZF_MAX_BLOCK_SIZE - fp->block_offset;
+    
+    while ( fp->gz_stream->avail_out != 0 ) {
+        // until we fill the output buffer (or hit EOF)
+        
+        if ( !input_eof && fp->gz_stream->avail_in == 0 ) {
+            // we are out of input data in the buffer. Get more.
             fp->gz_stream->next_in = fp->compressed_block;
+            fp->gz_stream->avail_in = hread(fp->fp, fp->compressed_block, BGZF_BLOCK_SIZE);
+            if ( fp->gz_stream->avail_in < 0 ) {
+                // hread had an error. Pass it on.
+                return fp->gz_stream->avail_in;
+            }
+            if ( fp->gz_stream->avail_in < BGZF_BLOCK_SIZE ) {
+                // we have reached EOF but the decompressor hasn't necessarily
+                input_eof = 1;
+            }
         }
-        else cached = 0;
-        do
-        {
-            fp->gz_stream->next_out = (Bytef*)fp->uncompressed_block + fp->block_offset;
-            fp->gz_stream->avail_out = BGZF_MAX_BLOCK_SIZE - fp->block_offset;
-            fp->gz_stream->msg = NULL;
-            ret = inflate(fp->gz_stream, Z_NO_FLUSH);
-            if (ret < 0 && ret != Z_BUF_ERROR) {
-                hts_log_error("Inflate operation failed: %s", bgzf_zerr(ret, ret == Z_DATA_ERROR ? fp->gz_stream : NULL));
-                fp->errcode |= BGZF_ERR_ZLIB;
-                return -1;
+        
+        fp->gz_stream->msg = NULL;
+        // decompress as much data as we can
+        int ret = inflate(fp->gz_stream, Z_SYNC_FLUSH);
+        
+        if ( (ret < 0 && ret != Z_BUF_ERROR) || ret == Z_NEED_DICT ) {
+            // an error occurred, other than running out of space
+            hts_log_error("Inflate operation failed: %s", bgzf_zerr(ret, ret == Z_DATA_ERROR ? fp->gz_stream : NULL));
+            fp->errcode |= BGZF_ERR_ZLIB;
+            return -1;
+        } else if ( ret == Z_STREAM_END ) {
+            // we finished a GZIP member
+            if ( input_eof ) {
+                // we are done with the last stream in the file
+                break;
+            } else if (fp->gz_stream->avail_in > 0) {
+                // try and read another GZIP member in the remaining data
+                int reset_ret = inflateReset(fp->gz_stream);
```

Comment:
I'm a little concerned about the case where the boundary of two gzip streams occurs precisely at the point where avail_in becomes zero.

However after some binary searching I have a test file that has this case and it works, despite not calling inflateReset.  The reset is necessary in other cases though, so this has my scratching my head.

Did you think about this case?

Reply:
Specfically `if (hpeek(fp->fp, &c, 1) == 1) {` instead of `avail_in > 0` will cause the reset to apply in this corner case too.

However I still don't quite understand why it's not needed.

Reply:
I did consider that case; the reason it works as is is that once `inflate` starts returning `Z_STREAM_END`, it will continue to return `Z_STREAM_END` when called again, until you `inflateReset` it.

So when the gzip stream ends, we call `inflateReset` if we have more data available right now, and skip it if we don't. Then the next time we go through, if there's more data, we'll put it in the input buffer, and when `inflate` returns `Z_STREAM_END` we'll see that we have more data available and reset it.

I don't think `hpeek(fp->fp, &c, 1) == 1` will work as a condition, because that doesn't catch the case when there's data in the input buffer but nothing more in the file. But `hpeek(fp->fp, &c, 1) == 1 || avail_in > 0` would probably work, and save a repeated `inflate` call.

Reply:
Aye realised I meant "> 0 || peek" shortly after posting, but then was in a meeting.   Thanks for the explanation as to how come it still worked.